vktrace/src/vktrace_layer/vktrace_lib_trace.cpp

```diff
@@ -136,6 +138,166 @@ void *strip_create_extensions(const void *pNext)
     return create_info;
 }
 
+#if defined (PLATFORM_LINUX)
+
+// When quering dirty pages from /proc/self/pagemap, there is a
+// small window between the reading of this file and when the dirty bits
+// are cleared in which some other thread might modify mapped memory.
+// We trap these memory writes by setting mapped memory pages to read-only and
+// setting the signal handler for SIGSEGV. The signal handler creates a list
+// of the addresses that caused a SIGSEGV. We keep the list in sighAddrList.
+// We used the semaphore sighAddrListSem to protect access to this list.
+
+static std::list<void *> sighAddrList;
+static vktrace_sem_id sighAddrListSem = NULL;
+
+static void segvHandler(int sig, siginfo_t *si, void *ununsed)
+{
+    size_t pageSize = getpagesize();
+
+    // Note: we use sem_wait and mprotect inside a signal handler.
+    // This is not POSIX compliant, but works on Linux.
+
+    // Add the addr that caused the segv to sighAddrList.
+    // It may not be in a mapped page we are tracking. If it
+    // isn't, we'll detect it later and generate an error then.
+    vktrace_sem_wait(sighAddrListSem);
+    sighAddrList.emplace_front(si->si_addr);
+    vktrace_sem_post(sighAddrListSem);
+
+    // Change protection of this page to allow the write to proceed
+    mprotect((void *)((uint64_t)si->si_addr & ~(pageSize-1)), pageSize, PROT_READ|PROT_WRITE);
```

Comment:
If this mprotect fails, the signal handler should call VKTRACE_FATAL_ERROR.
If the address cannot be made writable then returning from the signal handler will just retry the faulting instruction.
That will go right back into the signal handler again and again.